[patch] causes cargo run to succeed once, then fail, with no intervening changes

A somewhat pathological [patch] block can cause `cargo run` to succeed once, then fail on subsequent runs with no intervening changes to the crate, for a rather mind-bending experience.

Repro instructions:

1. Clone https://github.com/benesch/rust-play:

  ```
  $ git clone https://github.com/benesch/rust-play
  ```

2. Run `cargo run`, and watch it succeed:

  ```
  $ cd rust-play
  $ cargo run
      Updating git repository `https://github.com/benesch/timely-dataflow.git`...
     Compiling proc-macro2 v0.4.30
     Compiling ...
      Finished dev [unoptimized + debuginfo] target(s) in 1m 04s
       Running `target/debug/rust-play`
  ((0, (0, 0)), 0, 1)
  ((0, (0, 1)), 0, 1)
  ((1, (0, 2)), 0, 1)
  ((1, (0, 3)), 0, 1)
  ((2, (1, 4)), 0, 1)
  ((2, (1, 5)), 0, 1)
  ((3, (1, 6)), 0, 1)
  ((3, (1, 7)), 0, 1)
  ((4, (2, 8)), 0, 1)
  ((4, (2, 9)), 0, 1)
  ```

3. Run `cargo run` and watch it fail:

  ```
  $ cargo run
      Updating git repository `https://github.com/TimelyDataflow/timely-dataflow.git`
     Compiling timely_logging v0.10.0 (https://github.com/TimelyDataflow/timely-dataflow.git#b63bea65)
     Compiling timely_bytes v0.10.0 (https://github.com/TimelyDataflow/timely-dataflow.git#b63bea65)
     Compiling timely_communication v0.10.0 (https://github.com/TimelyDataflow/timely-dataflow.git#b63bea65)
     Compiling timely v0.10.0 (https://github.com/TimelyDataflow/timely-dataflow.git#b63bea65)
     Compiling differential-dataflow v0.10.0 (https://github.com/TimelyDataflow/differential-dataflow.git#c78b860a)
     Compiling rust-play v0.1.0 (/Users/benesch/Desktop/rust-play)
  error[E0412]: cannot find type `TestPatch` in module `timely`
    --> src/main.rs:38:22
     |
  38 | const _TEST: timely::TestPatch = timely::TestPatch;
     |                      ^^^^^^^^^ not found in `timely`
  
  error[E0425]: cannot find value `TestPatch` in module `timely`
    --> src/main.rs:38:42
     |
  38 | const _TEST: timely::TestPatch = timely::TestPatch;
     |                                          ^^^^^^^^^ not found in `timely`
  
  error: aborting due to 2 previous errors
  
  Some errors have detailed explanations: E0412, E0425.
  For more information about an error, try `rustc --explain E0412`.
  error: Could not compile `rust-play`.
  
  To learn more, run the command again with --verbose.
  ```

The reason it fails the second time around is because the first run leaves this [change to Cargo.lock lying around](https://gist.github.com/benesch/d6e38479f76b7a2cc61ff135df7d2bb3). Somehow the state that the first patch resolution writes causes the next patch resolution to fail.

Part of the problem here is that the `rust-play` crate [depends on `https://github.com/TimelyDataflow/timely-dataflow.git` (note the ".git" suffix) directly](https://github.com/benesch/rust-play/blob/6149e7277905e77e2bd2b9fc4670cae48ec197c3/Cargo.toml#L9), and [`https://github.com/TimelyDataflow/timely-dataflow` (no ".git" suffix) via differential dataflow](https://github.com/TimelyDataflow/differential-dataflow/blob/master/Cargo.toml#L32). I realize this might seem contrived here, but this actually happened accidentally in a real project, and it was extraordinarily confusing to debug.

I guess I don't understand whether it's intended that Cargo can determine when two different URLs actually point to the same crate. In `rust-play` without the `[patch]` block, i.e., with this diff applied

```diff
diff --git a/Cargo.toml b/Cargo.toml
index 732babc..0396948 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -8,5 +8,5 @@ edition = "2018"
 differential-dataflow = { git = "https://github.com/TimelyDataflow/differential-dataflow.git" }
 timely = { git = "https://github.com/TimelyDataflow/timely-dataflow.git" }
 
-[patch."https://github.com/TimelyDataflow/timely-dataflow"]
-timely = { git = "https://github.com/benesch/timely-dataflow.git", branch = "sample" }
+# [patch."https://github.com/TimelyDataflow/timely-dataflow"]
+# timely = { git = "https://github.com/benesch/timely-dataflow.git", branch = "sample" }
diff --git a/src/main.rs b/src/main.rs
index cec240f..712d8bc 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -35,4 +35,4 @@ fn main() {
 }
 
 // Comment this out to compile without the timely patch in Cargo.toml.
-const _TEST: timely::TestPatch = timely::TestPatch;
+// const _TEST: timely::TestPatch = timely::TestPatch;
```

Cargo somehow figures out that "timely-dataflow.git" and "timely" are actually the same crate, and uses one of them everywhere. The program wouldn't compile otherwise, since the types don't match in a way that causes compilation to fail, because differential exposes its copy of timely in its public API.

If it's intended that Cargo can handle having multiple URLs for the same crate, then it seems like there's just a bug in `[patch]` blocks where this same-crate-different-URL thing isn't getting sorted out. If Cargo isn't supposed to handle having multiple URLs for the same crate... then the situation is more complicated, and I'm not sure what the right fix is.

Anyway, I'm out of my depth here, but let me know if I can help at all!

I initially discovered this on macOS with Cargo 1.36, but it seems to reproduce readily with Cargo 1.37.

...actually, I just tried this on master, and it has a totally different presentation. Rather than failing to compile on the second try, it crashes outright:

```
$ RUST_BACKTRACE=1 ~/Sites/cargo/target/debug/cargo run
    Updating git repository `https://github.com/TimelyDataflow/timely-dataflow.git`
thread 'main' panicked at 'assertion failed: prev.is_none()', src/cargo/core/resolver/context.rs:153:25
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /Users/vsts/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /Users/vsts/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:47
   3: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:36
   4: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:200
   5: std::panicking::default_hook
             at src/libstd/panicking.rs:214
   6: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:477
   7: std::panicking::begin_panic
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b/src/libstd/panicking.rs:411
   8: cargo::core::resolver::context::Context::flag_activated
             at src/cargo/core/resolver/context.rs:153
   9: cargo::core::resolver::activate
             at src/cargo/core/resolver/mod.rs:655
  10: cargo::core::resolver::activate_deps_loop
             at src/cargo/core/resolver/mod.rs:382
  11: cargo::core::resolver::resolve
             at src/cargo/core/resolver/mod.rs:137
  12: cargo::ops::resolve::resolve_with_previous
             at src/cargo/ops/resolve.rs:338
  13: cargo::ops::resolve::resolve_with_registry
             at src/cargo/ops/resolve.rs:119
  14: cargo::ops::resolve::resolve_ws_with_opts
             at src/cargo/ops/resolve.rs:72
  15: cargo::ops::cargo_compile::compile_ws
             at src/cargo/ops/cargo_compile.rs:303
  16: cargo::ops::cargo_compile::compile_with_exec
             at src/cargo/ops/cargo_compile.rs:251
  17: cargo::ops::cargo_compile::compile
             at src/cargo/ops/cargo_compile.rs:240
  18: cargo::ops::cargo_run::run
             at src/cargo/ops/cargo_run.rs:73
  19: cargo::commands::run::exec
             at src/bin/cargo/commands/run.rs:75
  20: cargo::cli::execute_subcommand
             at src/bin/cargo/cli.rs:179
  21: cargo::cli::main
             at src/bin/cargo/cli.rs:94
  22: cargo::main
             at src/bin/cargo/main.rs:41
  23: std::rt::lang_start::{{closure}}
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b/src/libstd/rt.rs:64
  24: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:49
  25: std::panicking::try::do_call
             at src/libstd/panicking.rs:296
  26: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:82
  27: std::panicking::try
             at src/libstd/panicking.rs:275
  28: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  29: std::rt::lang_start_internal
             at src/libstd/rt.rs:48
  30: std::rt::lang_start
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b/src/libstd/rt.rs:64
  31: cargo::init_git_transports
```

The panic seems pretty obviously introduced in #7118, though I'm not familiar enough with the code to say whether #7118 moved closer or further to solving this overarching issue.

--- a/src/cargo/core/registry.rs
+++ b/src/cargo/core/registry.rs
@@ -9,7 +9,7 @@ use crate::core::PackageSet;
 use crate::core::{Dependency, PackageId, Source, SourceId, SourceMap, Summary};
 use crate::sources::config::SourceConfigMap;
 use crate::util::errors::{CargoResult, CargoResultExt};
-use crate::util::{profile, Config};
+use crate::util::{profile, CanonicalUrl, Config};
 
 /// Source of information about a group of packages.
 ///
@@ -75,9 +75,9 @@ pub struct PackageRegistry<'cfg> {
     yanked_whitelist: HashSet<PackageId>,
     source_config: SourceConfigMap<'cfg>,
 
-    patches: HashMap<Url, Vec<Summary>>,
+    patches: HashMap<CanonicalUrl, Vec<Summary>>,
     patches_locked: bool,
-    patches_available: HashMap<Url, Vec<PackageId>>,
+    patches_available: HashMap<CanonicalUrl, Vec<PackageId>>,
 }
 
 /// A map of all "locked packages" which is filled in when parsing a lock file
@@ -230,6 +230,8 @@ impl<'cfg> PackageRegistry<'cfg> {
     /// `query` until `lock_patches` is called below, which should be called
     /// once all patches have been added.
     pub fn patch(&mut self, url: &Url, deps: &[Dependency]) -> CargoResult<()> {
+        let canonical = CanonicalUrl::new(url)?;
+
         // First up we need to actually resolve each `deps` specification to
         // precisely one summary. We're not using the `query` method below as it
         // internally uses maps we're building up as part of this method
@@ -284,7 +286,7 @@ impl<'cfg> PackageRegistry<'cfg> {
                         url
                     )
                 }
-                if summary.package_id().source_id().url() == url {
+                if *summary.package_id().source_id().canonical_url() == canonical {
                     failure::bail!(
                         "patch for `{}` in `{}` points to the same source, but \
                          patches must point to different sources",
@@ -317,8 +319,8 @@ impl<'cfg> PackageRegistry<'cfg> {
         // `lock` method) and otherwise store the unlocked summaries in
         // `patches` to get locked in a future call to `lock_patches`.
         let ids = unlocked_summaries.iter().map(|s| s.package_id()).collect();
-        self.patches_available.insert(url.clone(), ids);
-        self.patches.insert(url.clone(), unlocked_summaries);
+        self.patches_available.insert(canonical.clone(), ids);
+        self.patches.insert(canonical, unlocked_summaries);
 
         Ok(())
     }
@@ -340,8 +342,11 @@ impl<'cfg> PackageRegistry<'cfg> {
         self.patches_locked = true;
     }
 
-    pub fn patches(&self) -> &HashMap<Url, Vec<Summary>> {
-        &self.patches
+    pub fn patches(&self) -> Vec<Summary> {
+        self.patches
+            .values()
+            .flat_map(|v| v.iter().cloned())
+            .collect()
     }
 
     fn load(&mut self, source_id: SourceId, kind: Kind) -> CargoResult<()> {
@@ -472,7 +477,7 @@ impl<'cfg> Registry for PackageRegistry<'cfg> {
             // This means that `dep.matches(..)` will always return false, when
             // what we really care about is the name/version match.
             let mut patches = Vec::<Summary>::new();
-            if let Some(extra) = self.patches.get(dep.source_id().url()) {
+            if let Some(extra) = self.patches.get(dep.source_id().canonical_url()) {
                 patches.extend(
                     extra
                         .iter()
@@ -605,7 +610,11 @@ impl<'cfg> Registry for PackageRegistry<'cfg> {
     }
 }
 
-fn lock(locked: &LockedMap, patches: &HashMap<Url, Vec<PackageId>>, summary: Summary) -> Summary {
+fn lock(
+    locked: &LockedMap,
+    patches: &HashMap<CanonicalUrl, Vec<PackageId>>,
+    summary: Summary,
+) -> Summary {
     let pair = locked
         .get(&summary.source_id())
         .and_then(|map| map.get(&*summary.name()))
@@ -669,7 +678,7 @@ fn lock(locked: &LockedMap, patches: &HashMap<Url, Vec<PackageId>>, summary: Sum
                 // map, and we see if `id` is contained in the list of patches
                 // for that url. If it is then this lock is still valid,
                 // otherwise the lock is no longer valid.
-                match patches.get(dep.source_id().url()) {
+                match patches.get(dep.source_id().canonical_url()) {
                     Some(list) => list.contains(&id),
                     None => false,
                 }

--- a/src/cargo/core/resolver/resolve.rs
+++ b/src/cargo/core/resolver/resolve.rs
@@ -3,8 +3,6 @@ use std::collections::{HashMap, HashSet};
 use std::fmt;
 use std::iter::FromIterator;
 
-use url::Url;
-
 use crate::core::dependency::Kind;
 use crate::core::{Dependency, PackageId, PackageIdSpec, Summary, Target};
 use crate::util::errors::CargoResult;
@@ -114,8 +112,8 @@ impl Resolve {
         self.graph.path_to_top(pkg)
     }
 
-    pub fn register_used_patches(&mut self, patches: &HashMap<Url, Vec<Summary>>) {
-        for summary in patches.values().flat_map(|v| v) {
+    pub fn register_used_patches(&mut self, patches: &[Summary]) {
+        for summary in patches {
             if self.iter().any(|id| id == summary.package_id()) {
                 continue;
             }

--- a/src/cargo/core/source/source_id.rs
+++ b/src/cargo/core/source/source_id.rs
@@ -15,10 +15,9 @@ use url::Url;
 
 use crate::core::PackageId;
 use crate::ops;
-use crate::sources::git;
 use crate::sources::DirectorySource;
 use crate::sources::{GitSource, PathSource, RegistrySource, CRATES_IO_INDEX};
-use crate::util::{CargoResult, Config, IntoUrl};
+use crate::util::{CanonicalUrl, CargoResult, Config, IntoUrl};
 
 lazy_static::lazy_static! {
     static ref SOURCE_ID_CACHE: Mutex<HashSet<&'static SourceIdInner>> = Mutex::new(HashSet::new());
@@ -34,8 +33,8 @@ pub struct SourceId {
 struct SourceIdInner {
     /// The source URL.
     url: Url,
-    /// The result of `git::canonicalize_url()` on `url` field.
-    canonical_url: Url,
+    /// The canonical version of the above url
+    canonical_url: CanonicalUrl,
     /// The source kind.
     kind: Kind,
     /// For example, the exact Git revision of the specified branch for a Git Source.
@@ -80,7 +79,7 @@ impl SourceId {
     fn new(kind: Kind, url: Url) -> CargoResult<SourceId> {
         let source_id = SourceId::wrap(SourceIdInner {
             kind,
-            canonical_url: git::canonicalize_url(&url)?,
+            canonical_url: CanonicalUrl::new(&url)?,
             url,
             precise: None,
             name: None,
@@ -216,7 +215,7 @@ impl SourceId {
         let url = config.get_registry_index(key)?;
         Ok(SourceId::wrap(SourceIdInner {
             kind: Kind::Registry,
-            canonical_url: git::canonicalize_url(&url)?,
+            canonical_url: CanonicalUrl::new(&url)?,
             url,
             precise: None,
             name: Some(key.to_string()),
@@ -226,6 +225,12 @@ impl SourceId {
     /// Gets this source URL.
     pub fn url(&self) -> &Url {
         &self.inner.url
+    }
+
+    /// Gets the canonical URL of this source, used for internal comparison
+    /// purposes.
+    pub fn canonical_url(&self) -> &CanonicalUrl {
+        &self.inner.canonical_url
     }
 
     pub fn display_index(self) -> String {
@@ -508,7 +513,7 @@ impl Hash for SourceId {
     fn hash<S: hash::Hasher>(&self, into: &mut S) {
         self.inner.kind.hash(into);
         match self.inner.kind {
-            Kind::Git(_) => self.inner.canonical_url.as_str().hash(into),
+            Kind::Git(_) => self.inner.canonical_url.hash(into),
             _ => self.inner.url.as_str().hash(into),
         }
     }

--- a/src/cargo/ops/resolve.rs
+++ b/src/cargo/ops/resolve.rs
@@ -343,7 +343,7 @@ pub fn resolve_with_previous<'cfg>(
         Some(ws.config()),
         ws.features().require(Feature::public_dependency()).is_ok(),
     )?;
-    resolved.register_used_patches(registry.patches());
+    resolved.register_used_patches(&registry.patches());
     if register_patches {
         // It would be good if this warning was more targeted and helpful
         // (such as showing close candidates that failed to match). However,

--- a/src/cargo/sources/git/mod.rs
+++ b/src/cargo/sources/git/mod.rs
@@ -1,4 +1,4 @@
-pub use self::source::{canonicalize_url, GitSource};
+pub use self::source::GitSource;
 pub use self::utils::{fetch, GitCheckout, GitDatabase, GitRemote, GitRevision};
 mod source;
 mod utils;

--- a/src/cargo/sources/git/source.rs
+++ b/src/cargo/sources/git/source.rs
@@ -27,7 +27,7 @@ impl<'cfg> GitSource<'cfg> {
         assert!(source_id.is_git(), "id is not git, id={}", source_id);
 
         let remote = GitRemote::new(source_id.url());
-        let ident = ident(source_id.url())?;
+        let ident = ident(&source_id);
 
         let reference = match source_id.precise() {
             Some(s) => GitReference::Rev(s.to_string()),
@@ -59,58 +59,17 @@ impl<'cfg> GitSource<'cfg> {
     }
 }
 
-fn ident(url: &Url) -> CargoResult<String> {
-    let url = canonicalize_url(url)?;
-    let ident = url
+fn ident(id: &SourceId) -> String {
+    let ident = id
+        .canonical_url()
+        .raw_canonicalized_url()
         .path_segments()
-        .and_then(|mut s| s.next_back())
+        .and_then(|s| s.rev().next())
         .unwrap_or("");
 
     let ident = if ident == "" { "_empty" } else { ident };
 
-    Ok(format!("{}-{}", ident, short_hash(&url)))
-}
-
-// Some hacks and heuristics for making equivalent URLs hash the same.
-pub fn canonicalize_url(url: &Url) -> CargoResult<Url> {
-    let mut url = url.clone();
-
-    // cannot-be-a-base-urls (e.g., `github.com:rust-lang-nursery/rustfmt.git`)
-    // are not supported.
-    if url.cannot_be_a_base() {
-        failure::bail!(
-            "invalid url `{}`: cannot-be-a-base-URLs are not supported",
-            url
-        )
-    }
-
-    // Strip a trailing slash.
-    if url.path().ends_with('/') {
-        url.path_segments_mut().unwrap().pop_if_empty();
-    }
-
-    // HACK: for GitHub URLs specifically, just lower-case
-    // everything. GitHub treats both the same, but they hash
-    // differently, and we're gonna be hashing them. This wants a more
-    // general solution, and also we're almost certainly not using the
-    // same case conversion rules that GitHub does. (See issue #84.)
-    if url.host_str() == Some("github.com") {
-        url.set_scheme("https").unwrap();
-        let path = url.path().to_lowercase();
-        url.set_path(&path);
-    }
-
-    // Repos can generally be accessed with or without `.git` extension.
-    let needs_chopping = url.path().ends_with(".git");
-    if needs_chopping {
-        let last = {
-            let last = url.path_segments().unwrap().next_back().unwrap();
-            last[..last.len() - 4].to_owned()
-        };
-        url.path_segments_mut().unwrap().pop().push(&last);
-    }
-
-    Ok(url)
+    format!("{}-{}", ident, short_hash(id.canonical_url()))
 }
 
 impl<'cfg> Debug for GitSource<'cfg> {
@@ -241,56 +200,54 @@ impl<'cfg> Source for GitSource<'cfg> {
 #[cfg(test)]
 mod test {
     use super::ident;
+    use crate::core::{GitReference, SourceId};
     use crate::util::IntoUrl;
-    use url::Url;
 
     #[test]
     pub fn test_url_to_path_ident_with_path() {
-        let ident = ident(&url("https://github.com/carlhuda/cargo")).unwrap();
+        let ident = ident(&src("https://github.com/carlhuda/cargo"));
         assert!(ident.starts_with("cargo-"));
     }
 
     #[test]
     pub fn test_url_to_path_ident_without_path() {
-        let ident = ident(&url("https://github.com")).unwrap();
+        let ident = ident(&src("https://github.com"));
         assert!(ident.starts_with("_empty-"));
     }
 
     #[test]
     fn test_canonicalize_idents_by_stripping_trailing_url_slash() {
-        let ident1 = ident(&url("https://github.com/PistonDevelopers/piston/")).unwrap();
-        let ident2 = ident(&url("https://github.com/PistonDevelopers/piston")).unwrap();
+        let ident1 = ident(&src("https://github.com/PistonDevelopers/piston/"));
+        let ident2 = ident(&src("https://github.com/PistonDevelopers/piston"));
         assert_eq!(ident1, ident2);
     }
 
     #[test]
     fn test_canonicalize_idents_by_lowercasing_github_urls() {
-        let ident1 = ident(&url("https://github.com/PistonDevelopers/piston")).unwrap();
-        let ident2 = ident(&url("https://github.com/pistondevelopers/piston")).unwrap();
+        let ident1 = ident(&src("https://github.com/PistonDevelopers/piston"));
+        let ident2 = ident(&src("https://github.com/pistondevelopers/piston"));
         assert_eq!(ident1, ident2);
     }
 
     #[test]
     fn test_canonicalize_idents_by_stripping_dot_git() {
-        let ident1 = ident(&url("https://github.com/PistonDevelopers/piston")).unwrap();
-        let ident2 = ident(&url("https://github.com/PistonDevelopers/piston.git")).unwrap();
+        let ident1 = ident(&src("https://github.com/PistonDevelopers/piston"));
+        let ident2 = ident(&src("https://github.com/PistonDevelopers/piston.git"));
         assert_eq!(ident1, ident2);
     }
 
     #[test]
     fn test_canonicalize_idents_different_protocols() {
-        let ident1 = ident(&url("https://github.com/PistonDevelopers/piston")).unwrap();
-        let ident2 = ident(&url("git://github.com/PistonDevelopers/piston")).unwrap();
+        let ident1 = ident(&src("https://github.com/PistonDevelopers/piston"));
+        let ident2 = ident(&src("git://github.com/PistonDevelopers/piston"));
         assert_eq!(ident1, ident2);
     }
 
-    #[test]
-    fn test_canonicalize_cannot_be_a_base_urls() {
-        assert!(ident(&url("github.com:PistonDevelopers/piston")).is_err());
-        assert!(ident(&url("google.com:PistonDevelopers/piston")).is_err());
-    }
-
-    fn url(s: &str) -> Url {
-        s.into_url().unwrap()
+    fn src(s: &str) -> SourceId {
+        SourceId::for_git(
+            &s.into_url().unwrap(),
+            GitReference::Branch("master".to_string()),
+        )
+        .unwrap()
     }
 }

--- a/src/cargo/util/canonical_url.rs
+++ b/src/cargo/util/canonical_url.rs
@@ -1,0 +1,73 @@
+use crate::util::errors::CargoResult;
+use std::hash::{self, Hash};
+use url::Url;
+
+/// A newtype wrapper around `Url` which represents a "canonical" version of an
+/// original URL.
+///
+/// A "canonical" url is only intended for internal comparison purposes in
+/// Cargo. It's to help paper over mistakes such as depending on
+/// `github.com/foo/bar` vs `github.com/foo/bar.git`. This is **only** for
+/// internal purposes within Cargo and provides no means to actually read the
+/// underlying string value of the `Url` it contains. This is intentional,
+/// because all fetching should still happen within the context of the original
+/// URL.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub struct CanonicalUrl(Url);
+
+impl CanonicalUrl {
+    pub fn new(url: &Url) -> CargoResult<CanonicalUrl> {
+        let mut url = url.clone();
+
+        // cannot-be-a-base-urls (e.g., `github.com:rust-lang-nursery/rustfmt.git`)
+        // are not supported.
+        if url.cannot_be_a_base() {
+            failure::bail!(
+                "invalid url `{}`: cannot-be-a-base-URLs are not supported",
+                url
+            )
+        }
+
+        // Strip a trailing slash.
+        if url.path().ends_with('/') {
+            url.path_segments_mut().unwrap().pop_if_empty();
+        }
+
+        // For GitHub URLs specifically, just lower-case everything. GitHub
+        // treats both the same, but they hash differently, and we're gonna be
+        // hashing them. This wants a more general solution, and also we're
+        // almost certainly not using the same case conversion rules that GitHub
+        // does. (See issue #84)
+        if url.host_str() == Some("github.com") {
+            url.set_scheme("https").unwrap();
+            let path = url.path().to_lowercase();
+            url.set_path(&path);
+        }
+
+        // Repos can generally be accessed with or without `.git` extension.
+        let needs_chopping = url.path().ends_with(".git");
+        if needs_chopping {
+            let last = {
+                let last = url.path_segments().unwrap().next_back().unwrap();
+                last[..last.len() - 4].to_owned()
+            };
+            url.path_segments_mut().unwrap().pop().push(&last);
+        }
+
+        Ok(CanonicalUrl(url))
+    }
+
+    /// Returns the raw canonicalized URL, although beware that this should
+    /// never be used/displayed/etc, it should only be used for internal data
+    /// structures and hashes and such.
+    pub fn raw_canonicalized_url(&self) -> &Url {
+        &self.0
+    }
+}
+
+// See comment in `source_id.rs` for why we explicitly use `as_str()` here.
+impl Hash for CanonicalUrl {
+    fn hash<S: hash::Hasher>(&self, into: &mut S) {
+        self.0.as_str().hash(into);
+    }
+}

--- a/src/cargo/util/mod.rs
+++ b/src/cargo/util/mod.rs
@@ -1,5 +1,6 @@
 use std::time::Duration;
 
+pub use self::canonical_url::CanonicalUrl;
 pub use self::cfg::{Cfg, CfgExpr};
 pub use self::config::{homedir, Config, ConfigValue};
 pub use self::dependency_queue::DependencyQueue;
@@ -28,6 +29,7 @@ pub use self::workspace::{
     print_available_tests,
 };
 
+mod canonical_url;
 mod cfg;
 pub mod command_prelude;
 pub mod config;

--- a/tests/testsuite/patch.rs
+++ b/tests/testsuite/patch.rs
@@ -1338,3 +1338,68 @@ version. [..]
         )
         .run();
 }
+
+#[cargo_test]
+fn canonicalize_a_bunch() {
+    let base = git::repo(&paths::root().join("base"))
+        .file("Cargo.toml", &basic_manifest("base", "0.1.0"))
+        .file("src/lib.rs", "")
+        .build();
+
+    let intermediate = git::repo(&paths::root().join("intermediate"))
+        .file(
+            "Cargo.toml",
+            &format!(
+                r#"
+                    [package]
+                    name = "intermediate"
+                    version = "0.1.0"
+
+                    [dependencies]
+                    # Note the lack of trailing slash
+                    base = {{ git = '{}' }}
+                "#,
+                base.url(),
+            ),
+        )
+        .file("src/lib.rs", "pub fn f() { base::f() }")
+        .build();
+
+    let newbase = git::repo(&paths::root().join("newbase"))
+        .file("Cargo.toml", &basic_manifest("base", "0.1.0"))
+        .file("src/lib.rs", "pub fn f() {}")
+        .build();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            &format!(
+                r#"
+                    [package]
+                    name = "foo"
+                    version = "0.0.1"
+
+                    [dependencies]
+                    # Note the trailing slashes
+                    base = {{ git = '{base}/' }}
+                    intermediate = {{ git = '{intermediate}/' }}
+
+                    [patch.'{base}'] # Note the lack of trailing slash
+                    base = {{ git = '{newbase}' }}
+                "#,
+                base = base.url(),
+                intermediate = intermediate.url(),
+                newbase = newbase.url(),
+            ),
+        )
+        .file("src/lib.rs", "pub fn a() { base::f(); intermediate::f() }")
+        .build();
+
+    // Once to make sure it actually works
+    p.cargo("build").run();
+
+    // Then a few more times for good measure to ensure no weird warnings about
+    // `[patch]` are printed.
+    p.cargo("build").with_stderr("[FINISHED] [..]").run();
+    p.cargo("build").with_stderr("[FINISHED] [..]").run();
+}

